baseline-error-prone/src/test/java/com/palantir/baseline/errorprone/Slf4jConstantLogMessageTests.java

```diff
@@ -121,6 +122,22 @@ public void negative() {
                         "    log.info(new DummyMarker(), \"constant\");",
                         "    log.warn(new DummyMarker(), \"constant\");",
                         "    log.error(new DummyMarker(), \"constant\");",
+                        "}",
+                        "",
+                        "void f(String param) {",
+                        "    // Logging of a non-constant, but with no arguments so it's a message",
+                        "    log.trace(\"constant\" + param);",
```

Comment:
This is inefficient code which results in a great deal of overhead in hot paths! When `TRACE` level is disabled, `log.trace("constant{}", param);` is a simple volatile read (level check) where `log.trace("constant" + param);` evaluates the string concatenation before the level check!

Reply:
You're absolutely right about this being terrible code for performance reasons. Maybe I should take a step back and ask what the goal of this bug checker is?

I was looking at it as a way to detect incorrect construction of an SLF4J format at runtime: especially, mixing of string concatenation and passing arguments. So, correctness, rather than performance. For that use case, this is fine.

Would it be fair to say you see this as a more general enforcement that also requires that the logging call is very cheap, if it doesn't result in output?

If so, I'll decline this PR: it (intentionally!) allows inefficient code through.

Either way, the example you give is a perfect candidate for adding to this file's `@BugPattern(explanation="...")`, for more context on why the code it catches is a bug.

Reply:
> You're absolutely right about this being terrible code for performance reasons. Maybe I should take a step back and ask what the goal of this bug checker is?

The goal is to ensure that only constant-string messages are logged, as the description states ;-)
It's light on details describing _why_ that's the case, as you have highlighted. Performance is one reason, but also due to the unique way we handle log data, separating user-data from internal diagnostic info. Fully describing these in the error-prone check would be difficult because the reasons don't all apply broadly, although I'm of the opinion that the string component should _always_ be a compile-time constant!

> I was looking at it as a way to detect incorrect construction of an SLF4J format at runtime: especially, mixing of string concatenation and passing arguments. So, correctness, rather than performance. For that use case, this is fine.

I think you'd actually want a separate check -- if the format values is a constant string value, you can read it, count the slf4j placeholders, and match them with non-throwable args to verify that data will be logged. This check doesn't count placeholders at all :-)
In practice we tend to log the raw format string and the parameters separately without any interpolation, thus we don't have such a check here.

Reply:
> I think you'd actually want a separate check -- if the format values is a constant string value, you can read it, count the slf4j placeholders, and match them with non-throwable args to verify that data will be logged. This check doesn't count placeholders at all :-)

We have such a check internally, and while I've made such a claim in the past, it looks like this time we might _actually_ be working towards open-sourcing all our Error Prone code. If you could wait for that (exact timeline TBD, but let's say 2-3 months) then this might save you some time :) (Our check also auto-fixes cases where the developer used `%s` instead of `{}`.)

Reply:
Thanks for the details; I think a separate check is exactly what I'm after here. For now, an internal copy with these changes covers my requirements, and I'll look to pick up either google/error-prone#565 or a separate project later.

The mismatched `{}` count would be ideal; I've seen a few cases where a final error argument is assumed to be a `Throwable`, but isn't.